Add no isolation mode to worker API

ghale opened the pull request:
Adds the ability to start work in an async thread with no isolation (i.e. different thread, same classloader).

eskatos commented on subprojects/workers/src/main/java/org/gradle/workers/internal/NoIsolationWorkerFactory.java:
```diff
@@ -0,0 +1,68 @@
+                    return buildOperationExecutor.call(new CallableBuildOperation<DefaultWorkResult>() {
+                        @Override
+                        public DefaultWorkResult call(BuildOperationContext context) {
+                            return DirectInstantiator.INSTANCE.newInstance(workerImplementationClass).execute(spec);
```
:o:

Shouldn't we at least serialize and deserialize `spec` before executing?
This would mean supporting injection then.
See our "Running user code" gdoc.

ghale replied:
After discussing with @eskatos, we've determined immutability of the spec is already taken care of.  I've added an integTest specifically for the no isolation case.

melix commented on subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java:
```diff
@@ -102,6 +104,20 @@ public DefaultWorkResult call() throws Exception {
+        switch(isolationMode) {
+            case AUTO:
+            case CLASSLOADER:
+                return workerInProcessFactory;
```
:x: can we rename the local variables, it's really confusing...

ghale replied:
renamed to isolatedClassloaderWorkerFactory, noIsolationWorkerFactory, and daemonWorkerFactory.

melix commented on subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java:
```diff
@@ -153,6 +170,40 @@ DaemonForkOptions getDaemonForkOptions(Class<?> actionClass, WorkerConfiguration
         return toDaemonOptions(actionClass, paramTypes, configuration.getForkOptions(), configuration.getClasspath());
     }
 
+    private void validateWorkerConfiguration(WorkerConfiguration configuration) {
```
:o: I wonder if we could move this code to `IsolationMode` itself. Like: `IsolationMode#isCompatibleWith(WorkerConfiguration config)`.

ghale replied:
I've created a followup story for this.

ghale commented on subprojects/workers/src/main/java/org/gradle/workers/IsolationMode.java:
```diff
@@ -19,22 +19,26 @@
- * Forking mode for workers.
+ * Isolation mode for workers.
  *
  * @since 3.5
```
Done.

ghale commented on subprojects/workers/src/main/java/org/gradle/workers/WorkerConfiguration.java:
```diff
@@ -53,16 +53,16 @@
+     * @return the forking mode for this worker, see {@link IsolationMode}, defaults to {@link IsolationMode#AUTO}
      */
-    ForkMode getForkMode();
+    IsolationMode getIsolationMode();
```
Done.

ghale commented on subprojects/workers/src/main/java/org/gradle/workers/IsolationMode.java:
```diff
@@ -19,22 +19,26 @@
+     * Don't attempt to isolate the work, use in-process workers.
      */
-    NEVER,
+    NONE,
```
I'm not opposed at all to changing these, but I'm not sure I like these better.  I suppose "SHARED" would be ok.

ghale commented on subprojects/workers/src/main/java/org/gradle/workers/IsolationMode.java:
```diff
@@ -19,22 +19,26 @@
+     * Isolate the work in it's own classloader, use in-process workers.
      */
-    ALWAYS;
+    CLASSLOADER,
```
I don't like this better.  "OWN" makes it less clear to me - this doesn't read well in english.  Perhaps "SHARED_CLASSLOADER" vs "ISOLATED_CLASSLOADER"?

ghale commented on subprojects/workers/src/main/java/org/gradle/workers/IsolationMode.java:
```diff
@@ -19,22 +19,26 @@
+    /**
+     * Isolate the work in a separate process, use out-of-process workers.
+     */
+    PROCESS
```
I'm pretty sure we won't want to tie up reusability into this boolean as there could be multiple combinations when we start to consider potentially having remote processes in the future - reusability would probably end up being a separate enum.  The potential remote case is probably a better comparison - I was thinking "PROCESS" vs "REMOTE", but maybe "EXTERNAL_PROCESS" vs "REMOTE_PROCESS" (future)?

ghale commented:
@melix I think I've addressed your blocking concerns.

adammurdoch commented on subprojects/workers/src/main/java/org/gradle/workers/IsolationMode.java:
```diff
@@ -19,22 +19,26 @@
+ * @since 4.0
  */
 @Incubating
-public enum ForkMode {
```
should leave `ForkMode` there until _after_ the Android plugin has switched to using `IsolationMode`.

ghale replied:
I've added this back.

adammurdoch commented on subprojects/workers/src/main/java/org/gradle/workers/WorkerConfiguration.java:
```diff
@@ -53,16 +53,20 @@
+     *
+     * @since 4.0
      */
-    ForkMode getForkMode();
```
we should leave this property here until _after_ the Android plugin has changed to `isolationMode`.

ghale replied:
I've added this back.

ghale commented:
@adammurdoch I've added ForkMode back until we are ready to remove it.  I've also added some test coverage to check that IsolationMode.NONE does not in fact isolate the worker.

eskatos commented:
🎉
